Mathlib/AlgebraicGeometry/EllipticCurve/Weierstrass.lean

```diff
@@ -390,77 +390,74 @@ variable {R} (j : R)
 
 /-- The Weierstrass curve $Y^2 + (j - 1728)XY = X^3 - 36(j - 1728)^3X - (j - 1728)^5$.
 It is of j-invariant j if it is an elliptic curve. -/
-def ofJ : WeierstrassCurve R :=
+def ofJNe0Or1728 : WeierstrassCurve R :=
   ⟨j - 1728, 0, 0, -36 * (j - 1728) ^ 3, -(j - 1728) ^ 5⟩
 
-lemma ofJ_c₄ : (ofJ j).c₄ = j * (j - 1728) ^ 3 := by
-  simp only [ofJ, c₄, b₂, b₄]
+lemma ofJNe0Or1728_c₄ : (ofJNe0Or1728 j).c₄ = j * (j - 1728) ^ 3 := by
+  simp only [ofJNe0Or1728, c₄, b₂, b₄]
   ring1
 
-lemma ofJ_Δ : (ofJ j).Δ = j ^ 2 * (j - 1728) ^ 9 := by
-  simp only [ofJ, Δ, b₂, b₄, b₆, b₈]
+lemma ofJNe0Or1728_Δ : (ofJNe0Or1728 j).Δ = j ^ 2 * (j - 1728) ^ 9 := by
+  simp only [ofJNe0Or1728, Δ, b₂, b₄, b₆, b₈]
   ring1
 
 end ModelsWithJ
 
-end WeierstrassCurve
-
 /-! ## Elliptic curves -/
 
-/-- An elliptic curve over a commutative ring. Note that this definition is only mathematically
+-- TODO: change to `protected abbrev IsElliptic := IsUnit W.Δ` once #17458 is merged
+/-- `WeierstrassCurve.IsElliptic` is a typeclass which asserts that a Weierstrass curve is an
+elliptic curve: that its discriminant is a unit. Note that this definition is only mathematically
 accurate for certain rings whose Picard group has trivial 12-torsion, such as a field or a PID. -/
-structure EllipticCurve (R : Type u) [CommRing R] extends WeierstrassCurve R where
-  /-- The discriminant `Δ'` of an elliptic curve over `R`, which is given as a unit in `R`. -/
-  Δ' : Rˣ
-  /-- The discriminant of `E` is equal to the discriminant of `E` as a Weierstrass curve. -/
-  coe_Δ' : Δ' = toWeierstrassCurve.Δ
+@[mk_iff]
+protected class IsElliptic : Prop where
+  isUnit : IsUnit W.Δ
 
-namespace EllipticCurve
+variable [W.IsElliptic]
 
-variable {R : Type u} [CommRing R]
+lemma isUnit_Δ : IsUnit W.Δ := IsElliptic.isUnit
 
-theorem toWeierstrassCurve_injective : Function.Injective (toWeierstrassCurve (R := R))
-  | ⟨x1, _, x3⟩, ⟨y1, _, y3⟩, h => by
-    change x1 = y1 at h
-    congr
-    exact Units.ext (by rw [x3, y3, h])
-
-@[ext]
-theorem ext {x y : EllipticCurve R} (h₁ : x.a₁ = y.a₁) (h₂ : x.a₂ = y.a₂) (h₃ : x.a₃ = y.a₃)
-    (h₄ : x.a₄ = y.a₄) (h₆ : x.a₆ = y.a₆) : x = y :=
-  toWeierstrassCurve_injective (WeierstrassCurve.ext h₁ h₂ h₃ h₄ h₆)
+/-- The discriminant `Δ'` of an elliptic curve over `R`, which is given as a unit in `R`.
+Note that to prove two equal elliptic curves have the same `Δ'`, you need to use `simp_rw`,
+as `rw` cannot transfer instance `WeierstrassCurve.IsElliptic` automatically. -/
+noncomputable def Δ' : Rˣ := W.isUnit_Δ.unit
```

Comment:
With the prime linter, I think we should try to rename this? What do you think of calling one of them `disc`, and keep the other one `Δ`? The one that's used more often should be `Δ`, but I'm not sure which.

Reply:
I don't know. Maybe we should leave this for another PR. Changing `Δ'` to something else requires altering some of existing code using `Δ'`.

Reply:
Sure!